core/store/src/trie/trie_storage.rs

```diff
@@ -1,33 +1,200 @@
 use std::cell::{Cell, RefCell};
 use std::io::ErrorKind;
 
-/// Wrapper over LruCache which doesn't hold too large elements.
+pub(crate) struct BoundedQueue<T> {
+    queue: VecDeque<T>,
+    /// If queue size exceeds capacity, item from the tail is removed.
+    capacity: usize,
+}
+
+impl<T> BoundedQueue<T> {
+    pub(crate) fn new(capacity: usize) -> Self {
+        // Reserve space for one extra element to simplify `put`.
+        Self { queue: VecDeque::with_capacity(capacity + 1), capacity }
+    }
+
+    pub(crate) fn clear(&mut self) {
+        self.queue.clear();
+    }
+
+    pub(crate) fn pop(&mut self) -> Option<T> {
+        self.queue.pop_front()
+    }
+
+    pub(crate) fn put(&mut self, key: T) -> Option<T> {
+        self.queue.push_back(key);
+        if self.queue.len() > self.capacity {
+            Some(self.pop().expect("Queue cannot be empty"))
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.queue.len()
+    }
+}
+
+/// In-memory cache for trie items - nodes and values. All nodes are stored in the LRU cache with three modifications.
+/// 1) Size of each value must not exceed `TRIE_LIMIT_CACHED_VALUE_SIZE`.
+/// Needed to avoid caching large values like contract codes.
+/// 2) If we put new value to LRU cache and total size of existing values exceeds `total_sizes_capacity`, we evict
+/// values from it until that is no longer the case. So the actual total size should never exceed
+/// `total_size_limit` + `TRIE_LIMIT_CACHED_VALUE_SIZE`.
+/// Needed because value sizes generally vary from 1 B to 500 B and we want to count cache size precisely.
+/// 3) If value is popped, it is put to the `deletions` queue with `deletions_queue_capacity` first. If popped value
+/// doesn't fit in the queue, the last value is removed from the queue and LRU cache, and newly popped value is inserted
+/// to the queue.
+/// Needed to delay deletions when we have forks. In such case, many blocks can share same parent, and we want to keep
+/// old nodes in cache for a while to process all new roots. For example, it helps to read old state root.
+pub struct SyncTrieCache {
+    /// LRU cache keeping mapping from keys to values.
+    cache: LruCache<CryptoHash, Arc<[u8]>>,
+    /// Queue of items which were popped, which postpones deletion of old nodes.
+    deletions: BoundedQueue<CryptoHash>,
+    /// Current total size of all values in the cache.
+    total_size: u64,
+    /// Upper bound for the total size.
+    total_size_limit: u64,
+    /// Shard id of the nodes being cached.
+    pub shard_id: u32,
+    /// Whether cache is used for view calls execution.
+    pub is_view: bool,
+}
+
+impl SyncTrieCache {
+    pub(crate) fn new(
+        cache_capacity: usize,
+        deletions_queue_capacity: usize,
+        total_size_limit: u64,
+        shard_id: u32,
+        is_view: bool,
+    ) -> Self {
+        Self {
+            cache: LruCache::new(cache_capacity),
```

Comment:
```suggestion
            // We use `total_size_limit` to bound the cache. 
            cache: LruCache::unbounded(),
```

Reply:
We actually need both:
- I don't want to think what happens if one decides to use this cache for storing many empty vectors
- this PR sets big total size limit on *all* caches to simplify configuration changes. So for all other shards - except 3 - we still need cache capacity limit

Reply:
makes sense